### PR TITLE
Add border images

### PR DESCRIPTION
**Pull request by TheAvidDev**
This PR adds support for attaching a set of 8 images onto the borders of containers, inspired by [@ascent12's comment](https://github.com/swaywm/sway/pull/4254#issuecomment-502368205) on an earlier iteration of this idea. This should allow for exterior rounded borders, (offset) drop shadows, and double borders, partially or fully resolving the following issues: #1065, #1426, #1511, and #3380.

---

The new command will most likely be used as follows in the config:
```
exec_always border-images.focused /some/folder/
exec_always border-images.focused_inactive /some/folder/
exec_always border-images.unfocused /some/folder/
exec_always border-images.urgent /some/folder/
```
This can't be a config time command because for a texture to be made, the renderer needs to be initialized. Possible solutions are delaying the running of the command until after the renderer has initialized, or saving the folder path in the config and initializing the actual changes later. I'm not really sure what's better, of if this is alright.

For some sample and testing files, feel free to checkout [this folder](https://github.com/TheAvidDev/sway-borders/tree/master/contrib/borders/).

**Edit**: For real life use of this, see [here](https://github.com/TheAvidDev/dotfiles#screenshots).

---

As an aside, I initially intended for this to be a separate package, but I found [this comment](https://github.com/swaywm/sway/issues/1065#issuecomment-275945423) which mentioned it may be accepted. Even if this is rejected, I would  be very thankful of any feedback received, so thanks in advance.

### CODE REVIEW

File: sway/desktop/render.c [hunk 1]
```diff
@@ -686,6 +686,173 @@ struct parent_data {
...
+
+	// TODO: Fix texture drawing between outputs.
+	float matrix[9];
+	memcpy(matrix, wlr_output->transform_matrix, sizeof(matrix));
+	wlr_matrix_translate(matrix, box.x, box.y);
+	wlr_matrix_scale(matrix, box.width, box.height);
+
+	pixman_region32_t texture_damage;
+	pixman_region32_init_rect(&texture_damage, box.x, box.y, box.width, box.height);
+	wlr_output_damage_add(output->damage, &texture_damage);
```
[the remark this answers is not in the document]
Comment (mortie) [reply sway/desktop/render.c:707]: So, I was looking at this stuff to fix it for myself, It seems like `apply_container_state` in `transaction.c` might be the right place to fix this; it deals with damaging the old location of a container, damages the new location, then changes the coordinates/dimensions of the container. I hacked together a fix by adding `- 40` to the X/Y locations and adding `+ 80` to the widths/heights.

I don't know if existing properties like `container->view->saved_geometry.x` or `container->saved_buffers[n].x` can be re-used, or if the `sway_container` struct needs additional properties, or if `transactions.c` should use the border_images values from `config`, or what.

**Comment by mortie** [comment]
There seems to be some artifacting with fractional scaling. There's a thin border around shadows on my 4k display scaled to 1.5x:
image

**Comment by markstos** [comment]
I like the idea of a custom border solution, but I wonder if this is the way to go about it.

I envision a config file command that allows me to set a filter command for windows. The filter command would take the window as input return and return an updated image as output, possibly with a border, drop shadows, or transparency. One implementation of a filter command might allow me to define some CSS which describes how a I might like a border or drop shadow to look. Perhaps in-memory the  filter command would take those CSS commands and essentially pre-render a border effect so an effect can be be very quickly applied to windows without recalculations.

I admittedly know little about how compositors work and if my idea is even workable with Wayland, but I'll say that managing custom borders through a folder of 8 static images is not appealing.

**Comment by TheAvidDev** [comment]
> I envision a config file command that allows me to set a filter command for windows.

This solution is meant to avoid exactly this -- 16 config options for everything you want, although that may be more intuitive.

> One implementation of a filter command might allow me to define some CSS which describes how a I might like a border or drop shadow to look.

There's nothing that can parse and render CSS in Sway at the moment and adding this would be a huge task. 

> I'll say that managing custom borders through a folder of 8 static images is not appealing.

While this may not be a huge improvement for you, this may become only a single image.

**Comment by markstos** [comment]
> This solution is meant to avoid exactly this -- 16 config options for everything you want, although that may be more intuitive.

I envisioned just one config option added to `sway`-- to specify a filter. Adding a drop shadow in CSS certainly doesn't take 16 configuration details either:

```
#example2 {
  box-shadow: 5px 10px #888888;
}
```
 
> There's nothing that can parse and render CSS in Sway at the moment and adding this would be a huge task.

I'm not suggesting that Sway learn how to parse CSS. I'm suggesting an external filter command, where one possible implementation would to be to use one that understands CSS. Another filter command might load a directory containing 8 static images.  Rust, Python and other languages have CSS parsing libraries.

**Comment by mortie** [comment]
> I'm not suggesting that Sway learn how to parse CSS. I'm suggesting an external filter command, where one possible implementation would to be to use one that understands CSS. Another filter command might load a directory containing 8 static images. Rust, Python and other languages have CSS parsing libraries.

I'm not sure I understand. What would the "filter" command do? Would it run once on startup/config change and deliver shadow images? Would it run once every time a window changes and return a pixel buffer

**Comment by kqzkqz** [comment]
perhaps related to the damage tracking issues, but I'm getting this when changing a floating window to tiled, and then manually positioning them under each other.

layout

**Comment by markstos** [comment]
>   I'm not sure I understand. 

@mortie To be clear, I'm not sure *I* understand enough about compositors enough to know if my idea is good or viable. Primarily, as a Sway user I'm expressing that it seems like there ought to be  a better way to declare the border style I'd like besides providing a directory of images. Perhaps I would feel differently if there was a companion tool to make it easy to generate such a directory of images.

**Comment by TheAvidDev** [comment]
> perhaps related to the damage tracking issues, but I'm getting this when changing a floating window to tiled, and then manually positioning them under each other.

No, this is actually an issue with which container is getting borders drawn. Can you run `swaymsg -t get_tree` and output the `representation` of that workspace. Feel free to remove the container on the right (in the image) so it's just the one with extra borders.

**Comment by ddevault** [comment]
Giving this the executive NACK. Feel free to keep working on it and incorporate it into a fork, but it's not desirable upstream.

**Comment by mortie** [comment]
@ddevault What do you feel about the concept of shadows and such in general? Is the NACK just for this particular approach? If so, what rough requirements do you have for an implementation of border shadows?

**Comment by ddevault** [comment]
This is a general NACK to any flashy features of this sort. I want sway to be simple and reliable going forward, not to perpetually grow new shiny things. We ought to focus on reliability, performance, and improvements to what we already have. Sway upstream needs to grow up from the cool new kid on the block into the reliable mainstay of the open-source desktop. This isn't the first feature to get NACKed for this reason, and it won't be the last.

And again, I would be supportive of a sway fork which adds these sorts of flashier features. I'm certain there'd be demand for it.

**Comment by TheAvidDev** [comment]
> We ought to focus on reliability, performance, and improvements to what we already have.

Understandable.

> Feel free to keep working on it and incorporate it into a fork, but it's not desirable upstream.

Yep, I'll do exactly this, as planned before.

File: sway/desktop/render.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheAvidDev) [reply sway/desktop/render.c:707]: Fixed

File: sway/sway.5.scd [hunk 2]
```diff
@@ -111,6 +111,29 @@ They are expected to be used with *bindsym* or at runtime through *swaymsg*(1).
 *border* toggle
 	Cycles through the available border styles.
 
+*border-images.<class>* <folder_path>
```
[the remark this answers is not in the document]
Comment (TheAvidDev) [reply sway/sway.5.scd:114]: Fixed.

File: contrib/split-border-image.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (TheAvidDev) [reply contrib/split-border-image.py:1]: This will be turned into a single image creation script.

**Pull request closed without merging**